### ISSUE TITLE
BlackBoxOptim seems to break Julia help functionality

### ISSUE THREAD

**Issue by jmgnve** [robertfeldt/BlackBoxOptim.jl#73]
The help functionality stops working when loading blackboxoptim alone in a new julia v0.6 session:

```julia
julia> using BlackBoxOptim

help?> println
search:ERROR: ArgumentError: abstract type is not a valid key for type Symbol
Stacktrace:
 [1] setindex!(::Dict{Symbol,Void}, ::Void, ::String) at .\dict.jl:414
 [2] push! at .\set.jl:35 [inlined]
 [3] unique_from(::Array{Any,1}, ::Array{Symbol,1}, ::Set{Symbol}, ::Int64) at .\set.jl:157
 [4] unique(::Array{Any,1}) at .\set.jl:140
 [5] accessible(::Module) at .\docs\utils.jl:362
 [6] completions at .\docs\utils.jl:367 [inlined]
 [7] repl_search(::Base.TTY, ::String) at .\docs\utils.jl:124

help?> median
search:ERROR: ArgumentError: abstract type is not a valid key for type Symbol
Stacktrace:
 [1] setindex!(::Dict{Symbol,Void}, ::Void, ::String) at .\dict.jl:414
 [2] push! at .\set.jl:35 [inlined]
 [3] unique_from(::Array{Any,1}, ::Array{Symbol,1}, ::Set{Symbol}, ::Int64) at .\set.jl:157
 [4] unique(::Array{Any,1}) at .\set.jl:140
 [5] accessible(::Module) at .\docs\utils.jl:362
 [6] completions at .\docs\utils.jl:367 [inlined]
 [7] repl_search(::Base.TTY, ::String) at .\docs\utils.jl:124
```

Very helpful package otherwise :)

**Comment by robertfeldt** [comment]
Thanks for checking out the package and sorry for this bug. I'm not really sure where to start looking into this. Any pointers would be appreciated. 

You have never seen this with some other package?
Are you running the latest BlackBoxOptim cloned from github (or if not which version)?

**Comment by jmgnve** [comment]
I have not seen this bug in other packages.

I have checked out blackboxoptim. This is what Pkg.status() shows me:

```
 - BenchmarkTools                0.0.8
 - BlackBoxOptim                 0.3.0+             master
 - CSV                           0.1.5
```

For some more information, look at the last discussion on Julia gitter (https://gitter.im/JuliaLang/julia). There is talk about type piracy or overload a function without importing it.

I am not a very good Julia programmer, so I am afraid I cannot help that much. But if there is anything I can do, please tell/direct me.

Does the bug happen on your computer as well?

**Comment by robertfeldt** [comment]
Thanks, I'm trying things out as we speak but it seems it does, yes. Thing is that this probably relates to the Compat package and our interaction with it so I'm not even sure where to start looking. Will check our "imports" as you say. Since the stacktrace "implicates" setindex! there might be something around or ParamsDict type. Can't promise to fix today due to other commitments but will check at least in coming weekend.

**Comment by jmgnve** [comment]
Thanks. It is no hurry. I will let you know if someone on the gitter channel gives a hint on how to solve the issue.

**Comment by robertfeldt** [comment]
I can replicate this on my end. I also tried making sure we import Base.setindex! before adding implementations for it but no difference. Hmm, a bit stumped right now, will have to think about this. If @alyst has some input please share. This does not happen on Julia 0.5, only on 0.6.

**Comment by alyst** [comment]
I also can reproduce that. I don't think it's related to `Base.setindex!()` because all the overloads we do are for `DictChain`, so shouldn't interfere with the `Base` functionality.
I suspect that's because of how
```julia
"""
<docstring>
"""
@compat abstract type sometype end
```
is processed. Looks like instead of associating the doc with "sometype", the doc system tries to use `abstract type`.

Have you checked it with the latest `@compat`?

**Comment by alyst** [comment]
Hmm, I tried removing `@compat` everywhere, but the problem remains.

**Comment by alyst** [comment]
The more minimal reproducible example is `unique([1, "a"])`. It fails after `using BlackBoxOptim`. Looks like some `Base.` overloads, but I cannot figure out yet, which ones.

**Comment by alyst** [comment]
In `unique([1, "a"])` example, for some reason, when `BlackBoxOptim` is loaded, `seenR` and `outR` collections of `unique_from()` are of the same type as `seen` (`Set{Int64}`) and `out` (`Vector{Int64}`), respectively, instead of being `Set{Any}` and `Vector{Any}`.
However, `R` is correctly inferred as `Any`, and from the command line `convert(Set{Any}, Set([1]))` works as expected.

**Comment by alyst** [comment]
More minimal example: `g(R, x) = convert(Set{R}, x)`.
Before `BlackBoxOptim`
```
julia> g(Any, Set([1]))
Set(Any[1])
```

After
```
julia> g(Any, Set([1]))
Set([1])
```

**Comment by robertfeldt** [comment]
We have a few converts in there but not a lot of use of Set. Hard one to crack.

**Comment by alyst** [comment]
Actually, for `h(R, x) = convert(Vector{R}, x)` the behaviour is the same.

**Comment by robertfeldt** [comment]
There seem to be multiple issues involved here. The latest minimal example (g func above) is fixed by making the Base.convert from keyword args to ParamsDict, i.e.:

```julia
Base.convert(::Type{Dict{Symbol,Any}}, kwargs::Vector{Any}) =
    Dict(Tuple{Symbol,Any}[(convert(Symbol, k), convert(Any,v)) for (k,v) in kwargs])
```

 explicit rather than via Base.convert. But this does not fix the problem with `unique` or the original bug in this issue.

**Comment by robertfeldt** [comment]
Actually, I was mistaken. The convert to Set{Any} is unaffected by the change I described above.

**Comment by alyst** [comment]
Yes, that `convert()` is definitely a "type piracy" case.
I have a patch that renames it into `kwargs2dict()`, but there's still something else that disrupts `Set{Any}` conversion.

**Comment by robertfeldt** [comment]
Yes, I did the same but still doesn't fix it. Hmm, really nasty. I don't see how any of the converts for our own types could have this effect so maybe this is caused by something else.

**Comment by jmgnve** [comment]
I just tested one thing:

```
using Distributions

?mean   # search for help on mean

using BlackBoxOptim

?mean  # search for help on mean
```

That works. I don't know if that information is useful though, but does it indicate some conflict between Distributions and BlackBoxOptim? What's really weird is that this does not work:

```
using Distributions

using BlackBoxOptim

?mean  # search for help on mean
```

**Comment by robertfeldt** [comment]
Wow, that is really unexpected. Even if there is some name clash how can the order make a difference? I guess the help system could be caching the lookup though but if I check for median instead for mean in your first sequence it still works.

**Comment by jmgnve** [comment]
I wonder who has implemented the documentation system in Julia, and whether this person/these persons could give a hint/help to solve the problem? Perhaps worthwhile to contact one of them?

**Comment by robertfeldt** [comment]
Yeah, maybe they can shed some light on it. But since we have found a "bug" as outlined by @alyst above it seems likely we are causing the problems that happen later (right now in the doc system but could be in other code that depends on the behavior that we somehow manages to trip up). If we don't make progress on this soon we might need to post on the Julia discourse or something.

**Comment by alyst** [comment]
Probably it makes sense to report the issue at JuliaLang/julia, but I would prefer first to condense an example further (i.e. reproduce the behaviour just by adding a few unrelated `Base.convert` methods to the clean Julia session without even loading a package).

**Comment by robertfeldt** [comment]
@jmgnve is this still a problem on 0.7 and 1.0? If not I'd like to close this issue.

**Comment by jmgnve** [comment]
It seems to work.
